Starting server from the MVP and connection improvements

caspernorrbin opened the pull request:
Adds the option to start a server from the menu. This server is run in a separate thread and works identical to a dedicated server. The host still communicates with the server through a connection, but locally. To accomplish this, the `MvpServer` has been given a header and `DedicatedServer` has been added to still allow the server to be standalone.

Connecting to a server has also been improved. There is now a connect function that is called explicitly, instead of connecting in `OnStart` like previously. If failed, the connection retries until it connects or a threshold is reached.

joelsiks commented on src/mvp/main.cpp:
```diff
@@ -107,17 +107,30 @@ void CreateStartMenu() {
     GameData::startMenu->AddMenuOption(
         UI::menu::MenuOption::CreateFromDerived(connectOption));
 
     // Currently does nothing...
```
Remove this.

joelsiks commented on src/mvp/main.cpp:
```diff
@@ -107,17 +107,30 @@ void CreateStartMenu() {
-                GameData::engine->SetAndGetScene(g_sStore);
+                bool connected = gameManager->ConnectToServer();
+                if (connected) {
+                    GameData::engine->SetAndGetDisplayLayout(g_dlStore);
+                    GameData::engine->SetAndGetScene(g_sStore);
```
If we assume we are never going back to the start menu again this is reasonable. Otherwise we should update the current store with the old value, like:

```c++
g_dlStore = GameData::engine->SetAndGetDisplayLayout(GameData::startMenu);
g_sStore = GameData::engine->SetAndGetScene(GameData::startMenuScene);
```

joelsiks replied:
We didn't think about this earlier but now is a good time to change this.

joelsiks commented on src/mvp/main.cpp:
```diff
@@ -107,17 +107,30 @@ void CreateStartMenu() {
+            if (args.pressed) {
+                bool connected = gameManager->StartAndConnectToServer();
+                if (connected) {
+                    GameData::engine->SetAndGetDisplayLayout(g_dlStore);
+                    GameData::engine->SetAndGetScene(g_sStore);
```
See above comment.

caspernorrbin commented:
I changed `SwapEngineLayers()` to actually swap layers and not only swap in the menu. This way, we can also use it to swap back to the game.

I also investigated the issues we (at least I) have been experiencing with clients sometimes not validating when connecting. It seems to be related to the client readying up too quickly, with the packet arriving at the server before the validation has been completed. To fix this (temporarily), I added a small (100ms) delay after connecting before continuing.

JoXW100 reviewed (commented):
When disconnecting the non-host client and then reconnecting with a new client, game stops moving the ships. The coins still update, and new ships can be created, but they don't move.

caspernorrbin commented:
> When disconnecting the non-host client and then reconnecting with a new client, game stops moving the ships. The coins still update, and new ships can be created, but they don't move.

Support for disconnecting and reconnecting is not being added with this pull request. You should find the behavior similar to doing the same on a dedicated server. I have another PR on the way which adds the ability to reconnect to games correctly

joelsiks reviewed (approved):
Good changes with swapping layers and good work over all!
